src/feedback_arc_set.rs

```diff
@@ -0,0 +1,426 @@
+        (
+            NodeIndex::new(e.source().index()),
+            NodeIndex::new(e.target().index()),
+        )
+    }));
+
+    g.edge_references()
+        .filter(move |e| node_seq[&e.source().index()] >= node_seq[&e.target().index()])
+}
+
+fn good_node_sequence(
+    edge_refs: impl Iterator<Item = (NodeIndex<usize>, NodeIndex<usize>)>,
+) -> HashMap<usize, usize> {
+    let mut nodes = FasNodeContainer { nodes: Vec::new() };
+    let mut buckets = Buckets {
+        sinks_or_isolated: NodeLinkedList::new(),
+        sources: NodeLinkedList::new(),
+        bidirectional_pve_dd: Vec::new(),
+        bidirectional_nve_dd: Vec::new(),
+    };
+    // Lookup of node indices from input graph to indices into `nodes`
+    let mut graph_ix_lookup = HashMap::new();
+
+    // Build node entries
+    for (from_g_ix, to_g_ix) in edge_refs {
+        let mut fas_node_entry = |g_ix: NodeIndex<usize>| -> FasNodeIndex {
+            match graph_ix_lookup.get(&g_ix) {
+                Some(fas_ix) => *fas_ix,
+                None => {
+                    let fas_ix = FasNodeIndex(nodes.nodes.len());
+
+                    nodes.nodes.push(LinkedListEntry::new(FasNode {
+                        graph_ix: g_ix,
+                        out_edges: Vec::new(),
+                        in_edges: Vec::new(),
+                        out_degree: 0,
+                        in_degree: 0,
+                    }));
+
+                    graph_ix_lookup.insert(g_ix, fas_ix);
+
+                    fas_ix
+                }
+            }
+        };
+
+        let from_fas_ix = fas_node_entry(from_g_ix);
+        let to_fas_ix = fas_node_entry(to_g_ix);
+
+        nodes[from_fas_ix].data().out_edges.push(to_fas_ix);
+        nodes[to_fas_ix].data().in_edges.push(from_fas_ix);
+    }
+
+    // Set initial in/out-degrees
+    for entry in nodes.nodes.iter_mut() {
+        let node = entry.data();
+        node.out_degree = node.out_edges.len();
+        node.in_degree = node.in_edges.len();
+    }
+
+    // Add nodes to initial lists
+    for i in 0..nodes.nodes.len() {
+        let fas_ix = FasNodeIndex(i);
+        buckets
+            .suitable_bucket(fas_ix, &mut nodes)
+            .push_front(fas_ix, &mut nodes);
+    }
+
+    let mut s_1 = VecDeque::new();
+    let mut s_2 = VecDeque::new();
+
+    loop {
+        let mut some_moved = false;
+
+        while let Some(sink_fas_ix) = buckets.sinks_or_isolated.pop(&mut nodes) {
+            some_moved = true;
+            buckets.update_neighbour_node_buckets(sink_fas_ix, &mut nodes);
+            s_2.push_front(nodes[sink_fas_ix].data().graph_ix);
+        }
+
+        while let Some(source_fas_ix) = buckets.sources.pop(&mut nodes) {
+            some_moved = true;
+            buckets.update_neighbour_node_buckets(source_fas_ix, &mut nodes);
+            s_1.push_back(nodes[source_fas_ix].data().graph_ix);
+        }
+
+        if let Some(list) = buckets
```

Comment:
This seems like it could bring the complexity up to potentially `|E|^2`, but in the benchmark tests at least, this isn't the case.

I had a much more complex implementation using a linked list-of-lists so that the highest bucket is more immediately available, but the benchmark results were the same, and it would be harder to maintain.

Reply:
Yeah, this seems ok.

I just tried adding a trim function at the end of `update_neighbour_node_buckets` that trims the empty lists at the end of `bidirectional_pve_dd` (since deltas increase one by one, the complexity gets amortized) and just got around a 1% decrease in the benchmarks.
Something similar could be done for the negative deltas by keeping track of the first non-empty list index, but it may complicate the implementation too much.